astropy/cosmology/funcs.py

```diff
@@ -253,3 +253,47 @@ def f(z):
         raise CosmologyError(f"Best guess z={res.x} is very close to the lower z limit {zmin}.\n"
                              "Try re-running with a different zmin.")
     return res.x
+
+
+def z_at_value(func, fval, zmin=1e-8, zmax=1000, ztol=1e-8, maxfun=500,
+               method='Brent', bracket=None, verbose=False):
+    """``scalar_z_at_value`` for vectorized inputs. Docstring set below."""
+    # `fval` can be a Quantity, which isn't (yet) compatible w/ `numpy.nditer`
+    # so we strip it of units for broadcasting and restore the units when
+    # passing the elements to `scalar_z_at_value`.
+    unit = getattr(fval, 'unit', 1)  # can be unitless
+
+    # bracket must be an object array (assumed to be correct) or a 'scalar'
+    # bracket: 2 or 3 elt sequence
+    if not isinstance(bracket, np.ndarray):  # 'scalar' bracket
+        if bracket is not None and len(bracket) not in (2, 3):
+            raise ValueError("`bracket` is not an object array "
+                             "nor a 2 (or 3) element sequence.")
+        else:  # munge bracket into a 1-elt object array
+            bracket = np.array([bracket, ()], dtype=object)[:1]
+    if bracket.dtype != np.object_:
+        raise TypeError(f"`bracket` has dtype {bracket.dtype}, not 'O'")
+
+    # make multi-dimensional iterator for all but `method`, `verbose`
+    # TODO: figure out 'reduce_ok', so scalar returns scalar
+    with np.nditer(
+        [np.asanyarray(fval), zmin, zmax, ztol, maxfun, bracket, None],
+        flags = ['refs_ok'],
+        op_flags = [*[['readonly']] * 6,  # ← inputs  output ↓
+                    ['writeonly', 'allocate', 'no_subtype']],
+        casting="no",
+    ) as it:
+        for fv, zmn, zmx, zt, mfe, brkt, zs in it:  # ← elt-wise unpack and eval ↓
+            zs[...] = scalar_z_at_value(func, fv * unit, zmin=zmn, zmax=zmx,
+                                        ztol=zt, maxfun=mfe, bracket=brkt.item(),
+                                        # not broadcasted
+                                        method=method, verbose=verbose)
+        # since bracket is an object array, the output will be too, so it is
+        # cast to the same type as the function value.
+        result = it.operands[-1].astype(fv.dtype)  # zs
+
+    # degrade to scalar if 0d array  # TODO: rm when 'reduce_ok' works
+    return result if result.shape != (1,) else result.item()
```

Comment:
Should be possible to avoid this; see above.

Reply:
Done. Thanks!